File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        // return recursion(n - 1, m - 1, grid);
        int[][] dp=new int[n][m];
        return memo(n - 1, m - 1, grid, dp);
        //tabulation
        
    }
    public int recursion(int n, int m, int[][] grid){
        //base case
        if(n < 0 || m < 0 || grid[n][m]==1)return 0;
        if(n==0 && m==0)return 1;
        int ans = recursion(n - 1, m,grid) + recursion(n, m - 1, grid);
        return ans;
    }
    
    public int memo(int n, int m, int[][] grid, int[][] dp){
        if(n < 0 || m < 0 || grid[n][m] == 1)return 0;
        if(n == 0 && m == 0)return dp[n][m] = 1;
        
        if(dp[n][m] !=0)return dp[n][m];
        
        int ans = memo(n - 1, m,grid, dp) + memo(n, m - 1, grid, dp);
        return dp[n][m] = ans;
    }
}

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int numberOfUniquePath(int i, int j, int n, int m, vector<vector<int>>& grid, vector<vector<int>> &dp){
        if(i <0 || j<0 || i>=n || j>=m || grid[i][j]==1)
            return 0;
            
        if(dp[i][j] != -1) return dp[i][j];
        if(i == n-1 && j == m-1){
            return 1;
        }
        
        int down = numberOfUniquePath(i+1, j, n , m, grid, dp);
        int right = numberOfUniquePath(i, j+1, n, m, grid, dp);
        
        dp[i][j] = down + right;
        return dp[i][j];
        
    }
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int n= obstacleGrid.size();
        int m= obstacleGrid[0].size();
        vector<vector<int>> dp(n+1, vector<int>(m+1, -1));
        return numberOfUniquePath(0, 0, n, m, obstacleGrid, dp);
    }
};
